README
P5NCI
=====

0.30 - Fri Jun 30 21:33:41 UTC 2006

P5NCI provides a means of loading shared libraries and calling their functions
from Perl without writing any glue code -- or compiling any XS, besides what
you compile when you install this module.

Version 0.30 adds the wonderful P5NCI::Declare with almost no syntax required
to bind to functions.

Version 0.20 adds support for sending and receiving opaque pointers to and from NCI functions.  Don't modify them.  They're opaque.

INSTALLATION

To install this module:

   $ perl Build.PL
   $ perl ./Build
   $ perl ./Build test
   $ sudo ./Build install

You can also use the Makefile, though I don't recommend it.

DEPENDENCIES

This module requires these other modules and libraries:

	DynaLoader (core, to run)

	Module::Builder      (to build)
	ExtUtils::CBuilder   (to build)

	Test::Simple    0.47 (to test)
	Test::Exception 0.19 (to test)

COPYRIGHT AND LICENCE

Copyright (c) 2004, 2006 chromatic.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 5.8.x itself, specifically Perl version 5.8.5.


Changes
Revision history for Perl extension P5NCI.

0.30 Fri Jun 30 21:37:32 UTC 2006
	- rewrote thunk parameter generator with Norman Nunley (now more complete)
	- added P5NCI::Declare

0.20  Tue May 9 2006
	- added pointer support

0.10  Sat May  6 22:54:15 UTC 2006
	- fixed float argument mangling bug on x86 (should fix other platforms)
	- attempted to reduce memory use with DynaLoader tricks
	- first version released to the CPAN

0.03  Fri Aug 27 18:05:16 UTC 2004  (chromatic)
	- improve function reference handling in XS

0.02  Fri Aug 27 00:08:59 UTC 2004  (chromatic)
	- allow simple string types
	- avoid unnecessary recompiling
	- improve generated XS (hopefully better compiler compatibility)

0.01  Fri Aug 20 03:19:37 UTC 2004  (chromatic)
	- original version released


src/nci_test.c
#include <math.h>
#include <stdio.h>

double double_double( double d );
float  double_float(  float  f );
int    double_int(    int    i );
short  double_short(  short  s );

char * change_string( char *p );
float  square_root(   float x );

int multiply_ints( int x, int y );

typedef struct _some_struct
{
	int    x;
	double y;
} some_struct;

some_struct* make_struct();
void set_x_value( some_struct *s, int x );
int  get_x_value( some_struct *s );
void free_struct( some_struct *s );

double double_double( double d )
{
	return d * 2.0;
}

float double_float( float f )
{
	return f * 2.0;
}

int double_int( int i )
{
	return i * 2;
}

short double_short( short s )
{
	return s * 2;
}

int multiply_ints( int x, int y )
{
	return x * y;
}

static char s[] = "X string\n";
char* change_string( char *p )
{
	s[0] = p[0];
	return s;
}

float square_root( float x )
{
	return sqrt( x );
}

some_struct* make_struct ()
{
	some_struct *s = malloc( sizeof( some_struct ) );
	return s;
}

void set_x_value ( some_struct *s, int x )
{
	s->x = x;
	return;
}

int  get_x_value ( some_struct *s )
{
	return s->x;
}

void free_struct ( some_struct *s )
{
	free( s );
	return;
}


lib/P5NCI.pm
package P5NCI;

use strict;
use warnings;

use vars '$VERSION';
$VERSION = '0.30';

use Carp 'croak';

use DynaLoader ();
DynaLoader::bootstrap( 'P5NCI', $VERSION );

my $nci_lib = $DynaLoader::dl_librefs[-1];

sub dl_load_flags { 0x01 }

sub add_path
{
	my $path = shift;
	unshift @DynaLoader::dl_library_path, $path;
}

sub load_func
{
	my ($lib, $name, $signature) = @_;
	my $thunk_name               = "nci_$signature";
	my $func                     = DynaLoader::dl_find_symbol( $lib, $name );
	my $xs_thunk_name            = "XS_P5NCI_$thunk_name";
	my $thunk_pointer            = DynaLoader::dl_find_symbol( $nci_lib,
	                                                           $xs_thunk_name );

	croak( "Don't understand NCI signature '$signature'\n" )
		unless $thunk_pointer;

	DynaLoader::dl_install_xsub( $thunk_name, $thunk_pointer )
		unless defined &$thunk_name;

	my $thunk_func = __PACKAGE__->can( $thunk_name );
	return build_thunk( $thunk_func, $func ) if $thunk_func;
}

sub build_thunk
{
	my ($thunk, $func) = @_;
	return sub { return $thunk->( $func, @_ ); };
}

BEGIN
{
	*find_lib = \&DynaLoader::dl_findfile;
	*load_lib = \&DynaLoader::dl_load_file;
}

1;
__END__

=head1 NAME

P5NCI - Perl extension for loading shared libraries and their functions

=head1 SYNOPSIS

  use P5NCI;

  # find and load a shared library in a cross-platform fashion
  my $library_path = P5NCI::find_lib( 'nci_test' );
  my $library      = P5NCI::load_lib( $library_path );

  # load a function from the shared library
  my $double_func  = P5NCI::load_func( $library, 'double_double', 'dd' );

  # now use it
  my $two_dot_oh   = $double_func->( 1.0 );

=head1 DESCRIPTION

P5NCI provides a bare-bones, stripped down, procedural interface to shared
libraries installed on your system.  It allows you to call functions in them
without writing or compiling any glue code.

I recommend using L<P5NCI::Library> as it has a nicer interface, but you can do
everything through here if you really want.

=head1 FUNCTIONS

=over

=item find_lib( $library_name )

Finds and returns the full path to a library on your particular platform given
its short name.  For example, on Unix, passing a C<$library_name> of
C<nci_test> will give you the full path to F<libnci_test.so>, if it's
installed.  On Windows and Mac OS X this should do the right thing as well.

=item load_lib( $library_full_path )

Given the full path to a library (as returned from C<find_lib()> or specified
on your own if you really don't care about cross-platform coding), loads the
library, if possible, and returns it in an opaque scalar that you really
oughtn't examine.

=item load_func( $library, $function_name, $signature )

Given an opaque library from C<load_lib()> in C<$library>, the name of a
function within that library in C<$function_name>, and the signature of that
function in C<$signature>, loads and returns a subroutine reference that allows
you to pass values to and return values from the function.

This function itself will throw an exception if you fail to provide both a
function name and its signature.  It will also throw an exception if it does
not understand the signature.  That might not be an error -- it doesn't
understand a lot of valid signatures yet.

Signatures are simple strings representing the types of the arguments the
function takes in their simplest forms.  For example, a function that takes two
ints and returns an int would have a signature of C<iii>.  A function that
takes two ints and returns nothing (or void) has a signature of C<vii>.  The
currently working signature items are:

=over

=item C<d>, a double

=item C<f>, a float

=item C<i>, an integer

=item C<p>, a pointer (read-only for now, so be careful)

=item C<s>, a short

=item C<t>, a string

=item C<v>, void (nothing), valid only as an output type, not an input type

=back

B<Note:> The signature list will definitely expand and may change in the
future.

=back

=head1 SEE ALSO

L<DynaLoader>, L<Inline::C>, L<perlxs>.

=head1 AUTHOR

chromatic, E<lt>chromatic at wgz dot orgE<gt>.

Based on Parrot's NCI by Dan Sugalski, Leo Toetsch, and a host of other people
including me (a little bit, here and there).

Thanks to Bill Ricker for documentation fixes and other suggestions.

Thanks to Norman Nunley for pair programming to help fix the generation code.

=head1 BUGS

No known bugs, though the signature list is currently pretty small in what it
supports and what it I<can> support.  Right now, you can only bind to C
functions that take zero to four arguments.  The XS code takes a long time to
compile as it is.

Hopefully this approach uses much less memory than the naive implementation,
though it depends on how well your platform manages shared libraries.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2004, 2006 chromatic.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 5.8.x.

=cut


lib/P5NCI/Declare.pm
package P5NCI::Declare;

use strict;
use warnings;

use vars '$VERSION';
$VERSION = '0.30';

use P5NCI::Library;

use Attribute::Handlers autotie => { '__CALLER__::NCI' => __PACKAGE__ };

my %libs;

sub import
{
	my $class = shift;
	$libs{ caller() }  = P5NCI::Library->new( @_ );
}

sub UNIVERSAL::NCI :ATTR
{
	my ($package, $symbol, $referent, $attr, $data) = @_;

	my $lib          = $libs{$package};
	my ($name, $sig) = @$data;
	my $function     = $lib->load_function( $name, $sig );
	*$symbol         = $function;
}

1;
__END__

=head1 NAME

P5NCI::Declare - declarative syntax for P5NCI bindings

=head1 SYNOPSIS

  use P5NCI::Declare library => 'shared_library';

  sub perl_function :NCI( c_function => 'vii' );

  perl_function( 101, 77 );

=head1 DESCRIPTION

C<P5NCI::Declare> allows you to bind Perl functions to C functions with
subroutine attributes.

When you C<use> this module, you I<must> pass: the key C<library> where the
value is the name of the shared library you want to load (following the normal
L<P5NCI> conventions).  You may pass an I<optional> C<path> key, where the
value is the path to the library.

To bind a Perl function name to a C function, use a subroutine declaration with
the C<NCI> attribute.  The attribute takes a pair where the key is the name of
the function and the value is its P5NCI signature.

=head1 AUTHOR

chromatic, E<lt>chromatic at wgz dot orgE<gt>

=head1 BUGS

No known bugs.

=head1 COPYRIGHT

Copyright (c) 2006, chromatic.  Some rights reserved.

This module is free software; you can use, redistribute, and modify it under
the same terms as Perl 5.8.


build_lib/P5NCI/Build.pm
package P5NCI::Build;

use strict;
use warnings;

use File::Spec::Functions;

use base 'Module::Build';

sub ACTION_build
{
	my $self = shift;

	$self->build_test_library();
	$self->build_xs_files();
	$self->SUPER::ACTION_build( @_ );
}

sub build_test_library
{
	my $self     = shift;
	require ExtUtils::CBuilder or die "ExtUtils::CBuilder not installed!\n";

	my $b        = ExtUtils::CBuilder->new();
	my $src_file = catfile( 'src', 'nci_test.c' );
	my $obj_file = $b->object_file( $src_file );
	$obj_file    =~ s/src.+(nci_test.*)/$1/;

	return unless $self->should_build( $obj_file, $src_file );

	my $compiled = $b->compile( source => $src_file );
	my $lib_file = $b->link(   objects => $compiled );
}

sub should_build
{
	my ($self, $dest, @sources) = @_;

	return 1 unless -e $dest;

	for my $dependency (@sources)
	{
		return 1 if -M $dependency < -M $dest;
	}

	return;
}

sub build_xs_files
{
	my $self = shift;

	require P5NCI::GenerateXS or die "Can't load XS generator!\n";

	my ($xs_file, $call_list) = map { catfile( 'lib', $_ ) }
		qw( P5NCI.xs call_list.txt );

	return unless $self->should_build(
		$xs_file, $call_list, $INC{'P5NCI/GenerateXS.pm'}
	);

	P5NCI::GenerateXS::generate_xs( $xs_file, $call_list );
}

1;


build_lib/P5NCI/GenerateXS.pm
package P5NCI::Sig;

sub new
{
	my ($class, %args) = @_;
	bless \%args, $class;
}

for my $accessor (qw( c_type type ))
{
	no strict 'refs';
	*{ $accessor } = sub { $_[0]->{$accessor} };
}

package P5NCI::GenerateXS;

use strict;
use warnings;

sub generate_xs
{
	my $out_file = shift;
	die "Usage: $0 <output_file.xs>\n" unless @_;

	open( my $out, '>', $out_file ) or die "Can't write $out_file: $!\n";

	write_header( $out );
	write_thunks( $out );
}

sub type_args
{
	return
	{
		d => P5NCI::Sig->new( type => 'd', c_type => 'double' ),
		i => P5NCI::Sig->new( type => 'i', c_type => 'int'    ),
		f => P5NCI::Sig->new( type => 'f', c_type => 'float'  ),
		p => P5NCI::Sig->new( type => 'p', c_type => 'void *' ),
		s => P5NCI::Sig->new( type => 's', c_type => 'short'  ),
		t => P5NCI::Sig->new( type => 't', c_type => 'char *' ),
		v => P5NCI::Sig->new( type => 'v', c_type => 'void'   ),
	};
};

sub write_thunks
{
	my $out = shift;

	my $args        = type_args();
	my $combination = generate_combinations( keys %$args );

	while (my $combo = $combination->())
	{
		# void makes no sense as anything other than return value or all args
		next if index( $combo, 'v', 2 ) > 0;

		my ($return, @args) = map { $args->{ $_ } } split('', $combo);
		my $func            = generate_function( "nci_$combo", $return, @args );
		print $out "\n", $func;
	}
}

sub generate_combinations
{
	my @possibilities = @_;

	return generate_iterator( 2, 5, \@possibilities );
}

sub generate_iterator
{
	my ($from, $to, $items) = @_;
	my @prefix              = (0) x $from;

	return sub
	{
		return if @prefix > $to;
		my $ret = join( '', map { $items->[ $_ ] } @prefix );

		# increment counter rightward
		my $i = 0;
		while ( ++$prefix[$i] > $#{ $items } )
		{
			$prefix[$i] = 0;
			$i++;
			if ( $i == @prefix )
			{
				@prefix = ( ( 0 ) x ( @prefix + 1 ) );
				last;
			}
		}

		return $ret;
	};
}

sub generate_function
{
	my ($signature, $return, @types) = @_;
	my $function = <<END_HERE;
%s
%s( c_func%s )
	SV* c_func
%sPREINIT:
	%s(*func)(%s);
CODE:
	func   = INT2PTR(%s(*)(%s), SvIV(c_func) );
END_HERE

	if ($return->type() eq 'v')
	{
		$function .= "\t(*func)(%s);\n";
	}
	else
	{
		$function .= <<END_HERE;
	RETVAL = (*func)(%s);
OUTPUT:
	RETVAL
END_HERE
	}

	my ($siglist, $insiglist, $arglist, $typelist) = get_type_lists( @types );
	return sprintf( $function, $return->c_type, $signature, $insiglist,
		$arglist, $return->c_type(), $typelist, $return->c_type(), $typelist,
		$siglist );
}

sub get_type_lists
{
	my (@siglist, @arglist, @typelist);

	if ($_[0]->type() eq 'v')
	{
		return ( '' ) x 3, ' ';
	}

	for my $type (@_)
	{
		my $var_name = $type->type() . @siglist;
		push @siglist, $var_name;
		push @typelist, $type->c_type();
		last if $type->type() eq 'v';
		push @arglist, "\t" . $type->c_type() . ' ' . $var_name;
	}

	my $siglist   = join( ', ', @siglist  );
	my $insiglist = ", $siglist";
	my $arglist   = join( "\n", @arglist, ''  );
	my $typelist  = join( ', ', @typelist );

	return ( $siglist, $insiglist, $arglist, $typelist );
}

sub write_header
{
	my $out = shift;

	print $out <<END_HERE;
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "ppport.h"

#ifdef newXS
	#undef newXS
	#define newXS ;
#endif

MODULE = P5NCI  PACKAGE = P5NCI
END_HERE
}

1;


lib/P5NCI/Library.pm
package P5NCI::Library;

use strict;
use warnings;

use P5NCI;

use vars '$VERSION';
$VERSION = '0.30';

use Carp 'croak';

sub new
{
	my ($self, %args) = @_;

	P5NCI::add_path( $args{path} ) if $args{path};

	croak "No library given\n" unless $args{library};
	my $libpath = P5NCI::find_lib( $args{library} );

	croak "No library found\n" unless $libpath;

	my $library = P5NCI::load_lib( $libpath );
	bless { %args, lib => $library }, $self;
}

sub package
{
	my $self = shift;
	$self->{package} ||= 'main';
}

sub load_function
{
	my ($self, $function, $signature) = @_;

	croak "No function given\n"  unless $function;
	croak "No signature given\n" unless $signature;

	return P5NCI::load_func( $self->{lib}, $function, $signature );
}

sub install_function
{
	my ($self, $name, $sig) = @_;
	my $function            = $self->load_function( $name, $sig);
	my $package             = $self->package();

	no strict 'refs';
	*{ $package . '::' . $name } = $function;
	return $function;

}

1;
__END__

=head1 NAME

P5NCI::Library - an OO library to the Native Calling Interface for Perl 5

=head1 SYNOPSIS

	use P5NCI::Library;

	# load a shared library
	my $lib = P5NCI::Library->new( library => 'nci_test' );

	# fetch a reference to a function in a shared libarry
	my $double_double = $lib->load_function( 'double_double', 'dd' );
	my $two_point_oh  = $double_double->( 1.0 );

	# load a shared library and associate it with a namespace
	my $lib = P5NCI::Library->new( library => 'nci_test', package => 'NCI' );

	# install a function from the shared library into the namespace
	$lib->install_function( 'double_int', 'ii' );
	my $two = NCI::double_int( 1 );

	# or call it from that namespace
	package NCI;

	my $six = double_int( 3 );

=head1 DESCRIPTION

P5NCI::Library provides an object-oriented way of loading shared libraries and
their functions, including installing them in specified namespaces.  This makes
it easy to call C functions from Perl without writing any messy glue code
yourself -- or having to compile any XS!

=head1 METHODS

=over 4

=item C<new( library => $library_name, [ package => $package_name ] )>

Loads the library named in C<$library_name> if possible and returns a new
P5NCI::Library object.

This will throw an exception if you fail to provide a library name or if it
cannot find a library of the given name.  Note that C<$library_name> should
ideally be the cross-platform library name.  For example, use C<nci_test>
instead of F<libnci_test.so> on Unix, F<nci_test.dll> on Windows, or
F<libnci_test.dylib> on Mac OS X.

The optional C<package> parameter sets the name of the package to which the
library object can install loaded functions directly.  You don't have to do
this.  If you don't specify a package name, it will default to C<main>.

=item C<package()>

Returns the name of the package into which the object can install shared
library functions.

=item C<load_function( $function_name, $signature )>

Attempts to load the function named C<$function_name> with the signature
C<$signature> from the shared library this object represents.  Signatures are
simple strings representing the types of the arguments the function takes in
their simplest forms.  For example, a function that takes two ints and returns
an int would have a signature of C<iii>.  A function that takes two ints and
returns nothing (or void) has a signature of C<vii>.  The current working
signature items are:

=over

=item C<i>, an integer

=item C<f>, a float

=item C<d>, a double

=item C<s>, a short

=item C<t>, a string

=item C<v>, void (nothing)

=back

B<Note:> The signature list will definitely expand and may change in the
future.

This function returns a Perl subroutine reference to the library function.
Call it as you would any other subroutine reference.  Note that it will throw
an exeption if you pass the wrong number of values.  It will probably segfault
or do horrible things if you pass the wrong type of values.  That's what
happens when you play with C.

This function itself will throw an exception if you fail to provide both a
function name and its signature.  It will also throw an exception if it does
not understand the signature.  That might not be an error -- it doesn't
understand a lot of valid signatures yet.

=item C<install_function( $function_name, $function_signature )>

Loads the function with the given name and signature from the library the
object represents and installs it into the namespace associated with the
object.  This will throw the same exceptions as does C<load_function> under the same circumstances.

It also returns the same subroutine reference.

=back

=head1 AUTHOR

chromatic, E<lt>chromatic at wgz dot orgE<gt>

=head1 BUGS

No known bugs.  Several known limitations.

=head1 COPYRIGHT

Copyright (c) 2004, 2006 chromatic.  All rights reserved.

This module is free software; you can use, redistribute, and modify it under
the same terms as Perl 5.8.x.
